Added restaurant card implementation

abutuc opened the pull request:
Closes #1274

Added the redesign implementation of the restaurant card. The card has as input parameters "restaurantName", "restaurantType", "menuItems", "isFavorite" and "onFavoriteToggle" callback function, which allows us to abstract the logic from the component. Additionally, this card returns a Generic Card whose child is the actual restaurant card content.

<img width="331" alt="Restaurant Card" src="https://github.com/user-attachments/assets/b88dcd62-39ff-49e8-b9cd-ca2440311649">

Note: the background color and the elevation property are being inherited from the Generic Card implementation, so when the Generic Card implementation has these issues fixed, the Restaurant Card should look like the mock version.

# Review checklist
-   [ ] Terms and conditions reflect the current change
-   [ ] Contains enough appropriate tests
-   [ ] If aimed at production, writes a new summary in `whatsnew/whatsnew-pt-PT`
-   [ ] Properly adds an entry in `changelog.md` with the change
-   [ ] If PR includes UI updates/additions, its description has screenshots
-   [ ] Behavior is as expected
-   [ ] Clean, well-structured code

vitormpp commented on packages/uni_ui/lib/cards/restaurant_card.dart:
```diff
@@ -0,0 +1,232 @@
+        break;
+      default:
+        iconType = cow;
+    }
```
According to previous discussions, `uni_app` should "send" the icons to the RestaurantCard component. Therefore, the icons should be passed as part of the MenuItem object.

vitormpp commented on packages/uni_ui/lib/cards/restaurant_card.dart:
```diff
@@ -0,0 +1,232 @@
+        );
+}
+
+class MenuItem {
```
To avoid introducing models in the `uni_ui` package, I suggest creating a MenuItem widget in uni_ui instead of defining a MenuItem model (that would take parameters like `icon` and `name`) and handle the display logic for each menu item. The `RestaurantCard` would then receive a list of `MenuItem` widgets, and uni_app would be responsible for creating and passing this list.
What do you think?

DGoiana replied:
I'm a bit skeptical and I do not like the idea of having models in the ui, so you got a point there. However, if I understood what you're saying, we'd have to create a model in the ui and follow the following path: **create in ui -> instantiate in app -> pass to ui**, which is a bit strange.
Either way, or that or passing all the information destructured.

vitormpp replied:
To clarify my point: instead of defining any model in the `uni_ui` package, we should create a widget that takes attributes like `icon` and `name` as parameters. The `RestaurantCard` would then receive a list of these widgets.

thePeras replied:
I agree with @vitormpp. MenuItem can be a widget and the restaurant card receives a list of MenuItem widgets

abutuc commented:
I added the requested changes. After performing the rebase the final look of the card component is:

<img width="379" alt="restaurant_card" src="https://github.com/user-attachments/assets/d9f39532-b44b-4fab-b06a-64a8dfdaee8f" />

I left the UniIcon implementation on the restaurant_card.dart file, but only for the favorite button icons (the others are being passed as arguments). Maybe in the future it would be nice to have a uni-ui global UniIcon widget so that we can use Duotone Phosphor Icons without needing to save the svg file.

I hope I addressed each suggested change and sorry for the delay!
@DGoiana  @thePeras  @vitormpp

abutuc commented:
> @abutuc good work here, the card looks pretty neat! Overall, the code looks good, only 2 things are itching my brain. First, in `RestaurantCardHeader` try to make the distance between the title and the top equal to the distance between the title and the gray line. Second, I do not understand the need to have `UniIcon`, can you clarify it?
<img width="361" alt="changed" src="https://github.com/user-attachments/assets/b9b66bee-5b6e-448f-aeb2-c135f1f34caf" />

@DGoiana what do you think? This is the best I could do to even out, I can't seem to reduce the upper space between the text and the top border of the container. I am not sure if anyone could help me out in this detail.
To make it visually more vertically centered, I just added some bottom padding to the container widget. I attempted to vertically center the Row widget and each child, but nothing worked!

Regarding the UniIcon part, I realized I didn't actually need it in the RestaurantCardHeader widget, so I removed it. But my point was, currently in order to use a Phosphor Duotone Icon I need to use that UniIcon logic, since if you search for it on the Phospor Icons site, it normally shows as "This weight is not yet supported" on the Flutter tab. With the UniIcon widget which @thePeras made, you can use the PhosphorIconsDuotone static icon provider. However, since this UniIcon widget is not yet in the uni_ui lib, whenever we need to use a Duotone icon, we either download the svg or replicate the code.
Here's an example of what happens when you use the Icon widget and just pass a PhosphorIconsDuotone icon:
<img width="360" alt="Captura de ecrã 2025-01-02, às 17 45 54" src="https://github.com/user-attachments/assets/3522289e-5129-4613-9f5c-def611f84107" />
Do you see how the icons are not in duotone but in a thin or light weight?

DGoiana commented:
@abutuc it looks great! Regarding the `UniIcon`, the class is being introduced in #1379. We can later add those heart icons into it.

thePeras reviewed (approved):
Nice Work!! ✅ 🛫
